### PR TITLE
:seedling: Unit tests for git client

### PR DESCRIPTION
**Pull request by naveensrinivasan**
- Add test functions to the Git client for testing repository initialization, commit listing, and search functionality

[clients/git/client_test.go]
- Add `createTestRepo` function to set up a test repository
- Add `TestInitRepo` to test the initialization of a repository
- Add `TestListCommits` to test the listing of commits
- Add `TestSearch` to test the search functionality

#### What kind of change does this PR introduce?

(Is it a bug fix, feature, docs update, something else?)

- [ ] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the new behavior (if this is a feature change)?**

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note

```

### CODE REVIEW

File: clients/git/client_test.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+					},
+					{
+						Path: "test.txt",
+					},
+				},
+				Hits: 2,
+			},
+		},
+		{
+			name: "Search with invalid query",
```
Comment (spencerschrock) [inline clients/git/client_test.go:189]: what makes this an invalid query? Or are we just expecting 0 results?

**Comment by spencerschrock** [comment]
@naveensrinivasan do you plan to continue this one?

**Comment by naveensrinivasan** [comment]
> @naveensrinivasan do you plan to continue this one?

Yes.

File: clients/git/client_test.go (hunk 1 above)
Comment (naveensrinivasan) [reply clients/git/client_test.go:182]: Yes, this is due to our anticipation of zero results.

File: clients/git/client_test.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			uri := fmt.Sprintf(test.uri, repoPath)
+
+			client := &Client{}
+			err := client.InitRepo(uri, test.commitSHA, test.commitDepth)
+			if test.expectedErr == "" {
+				if err != nil {
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply clients/git/client_test.go]: Thanks, Fixed it.

File: clients/git/client_test.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+			commitSHA:   "HEAD",
+			commitDepth: -1,
+			expectedErr: "negative commit depth",
+		},
+	}
+
+	repoPath := createTestRepo(t)
+	t.Cleanup(func() {
+		os.RemoveAll(repoPath) // nolint:errcheck
+	})
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply clients/git/client_test.go]: Thanks, Fixed it.

File: clients/git/client_test.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+			commitSHA:   "",
+			commitDepth: 1,
+			expectedErr: "repository does not exist",
+		},
+		{
+			name:        "InvalidCommitSha",
+			uri:         "file://%s",
+			commitSHA:   "invalid",
+			commitDepth: 1,
+		},
```
Comment (spencerschrock) [inline clients/git/client_test.go:108]: the fact that this didn't error might be a bug. It looks like `plumbing.NewHash` uses an empty hash on error.
https://github.com/ossf/scorecard/blob/9e719948b37672d569c15370a6d35cf574eb1e37/clients/git/client.go#L90-L99

So if we want to fail on invalid commit sha, we would need to check for hexadecimal ourselves.

File: clients/git/client_test.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	})
+
+	if err != nil {
+		t.Fatalf("Failed to make initial commit: %v", err)
+	}
+
+	t.Cleanup(func() {
+		os.RemoveAll(dir) // nolint:errcheck
+	})
```
Comment (spencerschrock) [inline clients/git/client_test.go:77]: nit: can we move this right under the `os.MkdirTemp` err checking? If we hit any of the `t.Fatalf` calls between there and here, we never schedule this cleanup.

File: clients/git/client_test.go (hunk 4 above)
Comment (naveensrinivasan) [reply clients/git/client_test.go:108]: I am going to remove this test case and create an issue.

Comment (naveensrinivasan) [reply clients/git/client_test.go:108]: https://github.com/ossf/scorecard/issues/3392

File: clients/git/client_test.go (hunk 5 above)
Comment (naveensrinivasan) [reply clients/git/client_test.go:77]: I agree. Thanks.

File: clients/gitlabrepo/contributors_test.go [hunk 6]
```diff
@@ -1,86 +0,0 @@
-// Copyright 2023 OpenSSF Scorecard Authors
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply clients/gitlabrepo/contributors_test.go:1]: Yes, It was unintentional. Thanks for catching it.

File: clients/git/client_test.go (hunk 1 above)
Comment (naveensrinivasan) [reply clients/git/client_test.go:182]: Yes, I have taken care of it.

**Review by spencerschrock — APPROVED** [review]
Thanks

**Pull request merged**